util/Midi/MIDI.gd

```diff
@@ -0,0 +1,272 @@
+	var _header_length = swap32(input_file.get_32())
+	var _format = swap16(input_file.get_16())
+	var track_chunks = swap16(input_file.get_16())
+	var _division = swap16(input_file.get_16())
+
+	# Read data from the MIDI File
+	for chunk in range(track_chunks):
+		#print("++New Track++")
+		# Read Track Header
+		var _track_id = swap32(input_file.get_32())
+		var _track_lenght = swap32(input_file.get_32())
+
+		var end_of_track = false
+		var previous_status = 0
+
+		tracks.append(MidiTrack.new())
+
+		while !input_file.eof_reached() && !end_of_track:
+			# All MIDI Events contain a timecode and a status byte
+			var status_time_delta = 0
+			var status = 0
+
+			# Read Timecode
+			status_time_delta = read_value(input_file)
+			status = input_file.get_8()
+
+			if status < 0x80:
+				status = previous_status
+				input_file.seek(input_file.get_position()-1)
+
+			# Match on the event.
+			match (status & 0xF0):
+				EventName.VoiceNoteOff:
+					previous_status = status
+					var _channel = status & 0x0F
+					var note_id = input_file.get_8()
+					var note_velocity = input_file.get_8()
+					tracks[chunk].events.append(MidiEvent.new(MidiEvent.NoteOff, note_id,
+						note_velocity, status_time_delta))
+
+				EventName.VoiceNoteOn:
+					previous_status = status
+					var _channel = status & 0x0F
+					var note_id = input_file.get_8()
+					var note_velocity = input_file.get_8()
+					if note_velocity == 0:
+						tracks[chunk].events.append(MidiEvent.new(MidiEvent.NoteOff, note_id,
+							note_velocity, status_time_delta))
+					else:
+						tracks[chunk].events.append(MidiEvent.new(MidiEvent.NoteOn, note_id,
+							note_velocity, status_time_delta))
+				EventName.VoiceAftertouch:
+					previous_status = status
+					var _channel = status & 0x0F
+					var _note_id = input_file.get_8()
+					var _note_velocity = input_file.get_8()
+					tracks[chunk].events.append(MidiEvent.new(MidiEvent.Other))
+				EventName.VoiceControlChange:
+					previous_status = status
+					var _channel = status & 0x0F
+					var _note_id = input_file.get_8()
+					var _note_velocity = input_file.get_8()
+					tracks[chunk].events.append(MidiEvent.new(MidiEvent.Other))
+				EventName.VoiceChannelPressure:
+					previous_status = status
+					var _channel = status & 0x0F
+					var _channel_pressure = input_file.get_8()
+					tracks[chunk].events.append(MidiEvent.new(MidiEvent.Other))
+				EventName.VoicePitchBend:
+					previous_status = status
+					var _channel = status & 0x0F
+					var _LS7B = input_file.get_8()
+					var _MS7B = input_file.get_8()
+					tracks[chunk].events.append(MidiEvent.new(MidiEvent.Other))
+				EventName.SystemExclusive:
+					if status == 0xF0:
+						#System Exclusive Message Begin
+						print("System Exclusive Begin: " + read_string(input_file,
+							read_value(input_file)))
+					if status == 0xF7:
+						print("System Exclusive End: " + read_string(input_file,
+							read_value(input_file)))
+					if status == 0xFF:
```

Comment:
Maybe make this a match, and also make an enum from these constants?

Reply:
okay